### ISSUE TITLE
Prepare your apps for iOS 14 Apple announced the new AppTrackingTransparency framework, which requires changes to your iOS apps. Implement the GMA SDK 7.64.0 (or later) and set up consent messaging to help prevent a significant loss in ad revenue

### ISSUE THREAD

**Issue by khoabuidev** [invertase/react-native-firebase#4107]
Hi, Today I received this message on google Admob "Prepare your apps for iOS 14 Apple announced the new AppTrackingTransparency framework, which requires changes to your iOS apps. Implement the GMA SDK 7.64.0 (or later) and set up consent messaging to help prevent a significant loss in ad revenue."  so what can I do now for it? many thanks!
<img width="925" alt="Screen Shot 2020-08-19 at 7 08 39 PM" src="https://user-images.githubusercontent.com/12908326/90633202-d5791f00-e24f-11ea-9402-2caa4dc3db3e.png">

## Issue

Describe your issue here

---

### Javascript

<details><summary>Click To Expand</summary>
<p>

#### `package.json`:

```json
# N/A
```

#### `firebase.json` for react-native-firebase v6:

```json
# N/A
```

</details>

### iOS

<details><summary>Click To Expand</summary>
<p>

#### `ios/Podfile`:

- [ ] I'm not using Pods
- [x] I'm using Pods and my Podfile looks like:

```ruby
# N/A
```

#### `AppDelegate.m`:

```objc
// N/A
```

</p>
</details>

---

### Android

<details><summary>Click To Expand</summary>
<p>

#### Have you converted to AndroidX?

- [ ] my application is an AndroidX application?
- [ ] I am using `android/gradle.settings` `jetifier=true` for Android compatibility?
- [ ] I am using the NPM package `jetifier` for react-native compatibility?

#### `android/build.gradle`:

```groovy
// N/A
```

#### `android/app/build.gradle`:

```groovy
// N/A
```

#### `android/settings.gradle`:

```groovy
// N/A
```

#### `MainApplication.java`:

```java
// N/A
```

#### `AndroidManifest.xml`:

```xml
<!-- N/A -->
```

</p>
</details>

---

## Environment

<details><summary>Click To Expand</summary>
<p>

**`react-native info` output:**

```
 OUTPUT GOES HERE
```

- **Platform that you're experiencing the issue on**:
  - [ ] iOS
  - [ ] Android
  - [ ] **iOS** but have not tested behavior on Android
  - [ ] **Android** but have not tested behavior on iOS
  - [ ] Both
- **`react-native-firebase` version you're using that has this issue:**
  - `e.g. 5.4.3`
- **`Firebase` module(s) you're using that has the issue:**
  - `e.g. Instance ID`
- **Are you using `TypeScript`?**
  - `Y/N` & `VERSION`

</p>
</details>

---

- 👉 Check out [`React Native Firebase`](https://twitter.com/rnfirebase) and [`Invertase`](https://twitter.com/invertaseio) on Twitter for updates on the library.

**Comment by andersonaddo** [comment]
Hmmmm
so I had a look at this:
https://developers.google.com/admob/ios/ios14

Looks like there are two options: Stick with IFDA (which Apple is trying to effectively kill), or move on to SKAdNetwork.
And it seems that most of the work will still be developer side work for either option you choose - you just have to add some things to your info.plist.
Some native code checks are recommended but not needed. But I suppose patch-package can help with that. 

I wonder if we should think about integrating this: https://developers.google.com/admob/ump/ios/quick-start

**Comment by khoabuidev** [comment]
> Hmmmm
> so I had a look at this:
> https://developers.google.com/admob/ios/ios14
> 
> Looks like there are two options: Stick with IDFA (which Apple is trying to effectively kill), or move on to SKAdNetwork.
> And it seems that most of the work will still be developer side work for either option you choose - you just have to add some things to your info.plist.
> Some native code checks are recommended but not needed. But I suppose patch-package can help with that.
> 
> I wonder if we should think about integrating this: https://developers.google.com/admob/ump/ios/quick-start

Thank @andersonaddo very much, I will try to apply this.

**Comment by andersonaddo** [comment]
Just to note:
The docs note that Google Mobile Ads v7.64.0 will be needed, but I *think* we use a lower version. I'm not sure.
Having a look at our tests files (which contains a compiled app): 

https://github.com/invertase/react-native-firebase/blob/fcf24478f9c889021afc4426a7016cfd3fd5d6cc/tests/ios/Podfile.lock#L14

https://github.com/invertase/react-native-firebase/blob/fcf24478f9c889021afc4426a7016cfd3fd5d6cc/tests/ios/Podfile.lock#L219

**Comment by mikehardy** [comment]
Good spot @andersonaddo 

They specify '~> 7.50' https://github.com/CocoaPods/Specs/blob/b513b219e9a15e2ab11b7ed93e69e0c7336a176a/Specs/0/3/5/Firebase/6.30.0/Firebase.podspec.json#L93

https://guides.cocoapods.org/using/the-podfile.html#specifying-pod-versions
 `~> Major.Minor' (but not patch!)  means that pod update should be allowed to bring in the 7.64 Google-Mobile-Ads-SDK

That's why the test Podfile.lock has 7.62 (must have been current the last time the lock file was updated) vs 7.50 which is specified

So it's achievable to test using current dependencies I think

**Comment by andersonaddo** [comment]
Great.
@khoabuidev please keep us updated about this. If it works, it might be worth mentioning it in our docs.

**Comment by khoabuidev** [comment]
I have update all dependencies in my project, remove Podfile.lock and Pods folder and run "pod install" . I see that it installing Google-Mobile-Ads-SDK (7.64.0), detail as below:

1. npm i -g npm-check-updates
2. npx npm-check-updates -u
3. rm -r node_modules
4. yarn install
5. cd ios
6. pod install
And there is the result
<img width="409" alt="Screen Shot 2020-08-20 at 1 35 06 PM" src="https://user-images.githubusercontent.com/12908326/90725983-2d189880-e2eb-11ea-9477-ff0d16ba4b4a.png">
FYI: @andersonaddo

**Comment by mikehardy** [comment]
Of interest, react-native-permissions has a pull request that should work now, but it requires Xcode 12beta toolchain to compile https://github.com/react-native-community/react-native-permissions/pull/501

And adding the right stuff to your plist 

```
<key>NSUserTrackingUsageDescription</key>
<string>This identifier will be used to deliver personalized ads to you.</string>

```

**Comment by mikehardy** [comment]
With the addition of that PR to react-native-permissions to handle the user consent, and the ability (proven) to use the latest version of the AdMobs SDK, I don't *think* there is anything more to do here. I'm going to close this but if I am wrong obviously we can reopen and discuss - not closing because it's not important just that I think it's resolved :-). Cheers

**Comment by khoabuidev** [comment]
> Of interest, react-native-permissions has a pull request that should work now, but it requires Xcode 12beta toolchain to compile [react-native-community/react-native-permissions#501](https://github.com/react-native-community/react-native-permissions/pull/501)
> 
> And adding the right stuff to your plist
> 
> ```
> <key>NSUserTrackingUsageDescription</key>
> <string>This identifier will be used to deliver personalized ads to you.</string>
> ```

Thanks so much @mikehardy mikehardy

**Comment by mikehardy** [comment]
The related react-native-permissions PR has been merged, if you get `react-native-permissions@2.2.0` (or greater) and have Xcode12 + ios14 + `pod update` (to get the new Mobile-Ads-SDK) you should be set. 

Any reports of success or failure are welcome, this is the leading edge of development now but will affect lots of users so the reports will be useful to your fellow developers!

**Comment by manuhook** [comment]
Do you think there's any chance react-native-firebase will support the **User Messaging Platform (UMP) SDK** (https://developers.google.com/admob/ump/ios/quick-start) which replaces the previous open source Consent SDK currently integrated to rn-firebase ?
It will be a great help to manage automatically both RGPD or CCPA consent **and** iOS 14 AppTrackingTransparency. Thanks !

**Comment by mikehardy** [comment]
@manuhook that does look like a nice way forward. We'd gladly take a PR to migrate to it

**Comment by rochapablo** [comment]
Besides adding

```
<key>NSUserTrackingUsageDescription</key>
<string>This identifier will be used to deliver personalized ads to you.</string>
```

To get the alert request, should I do something else?

I'm not getting errors but neither getting that alert request permission.

**Comment by rochapablo** [comment]
I thought that the request for permission was already implemented, but it isn't. Is that right? I'm, asking just to be sure. I did and now the alert it's working.

```
import { request, PERMISSIONS } from 'react-native-permissions';
request(PERMISSIONS.IOS.APP_TRACKING_TRANSPARENCY)
```

**Comment by mikehardy** [comment]
@rochapablo wrong repository - that's a react-native-permissions question, this library does nothing to check or request the permission

**Comment by AlekseyKutsko** [comment]
Hi there!
Should I call consent form(AdsConsent.showForm) on ios if I'll add AppTrackingTransparency?

**Comment by emilioheinz** [comment]
Is there any roadmap to automatically request this permissions when using something that requires the `APP_TRACKING_TRANSPARENCY`? Or the final approach is requesting this permission with React Native Permissions? 

Thank you.

**Comment by mikehardy** [comment]
I am not interested in merging functionality here that handles permissions. react-native-permissions exists for this very purpose, it should be used

**Comment by emilioheinz** [comment]
Yes I totally Agree!

**Comment by brianGammon** [comment]
Hi all, question on this topic... if I do have the latest Google-Mobile-Ads-SDK (7.67 as of now), but I do not implement the request for permission using something like react-native-permissions, what happens? Will adMob work, but default to non-personalized ads and no tracking?

**Comment by emilioheinz** [comment]
@brianGammon I think after January 2021 , or as apple says early next year, it will stop working.

“On iOS 14, iPadOS 14, and tvOS 14, apps will be required to receive user permission to track users across apps or websites owned by other companies, or to access the device’s advertising identifier,” Apple posted today in a developer update. “We are committed to ensuring users can choose whether or not they allow an app to track them. To give developers time to make necessary changes, apps will be required to obtain permission to track users starting early next year. More information, including an update to the App Store Review Guidelines, will follow this fall.”

font: https://www.forbes.com/sites/johnkoetsier/2020/09/03/idfa-stay-of-execution-apple-delays-new-ios-14-privacy-measures-until-2021/#162e185d569f
font: https://developer.apple.com/news/?id=hx9s63c5

**Comment by bd-arc** [comment]
It looks like there hasn't been any feedback from the trenches on the AdMob/ATT issue yet, so allow me to share some.

We've been trying to push one of our apps' update for weeks now. We had been using AdMob for months, and have recently implemented the ATT — kudos to everyone in this thread, your conversations have been really helpful for that 👍 

Since we're based in Europe, we also need to deal with a few GDPR specifics. For example, Apple asked us to present the ATT prompt **before** the one that AdMob requires in the EEA.

But, so far, each attempt to publish the app (5 and counting) has been met with the following answer:

> We're looking forward to completing the review of your app, but we need more information to continue. Specifically, we noticed that your app uses the AppTrackingTransparency framework, but we haven't been able to locate the relevant AppTrackingTransparency permission requests.

> While it is not required to implement AppTrackingTransparency at this time, we check to make sure the implementation is compliant with our guidelines when we detect the framework in an app.

And now for the crazy part: **no device running the latest iOS version (14.2 at the time of this writing) seems to trigger the ATT prompt whereas every single XCode 12 simulator running exact the same version does!**

So it looks like developers are stuck in a loophole:

- we're required to implement the ATT
- apps are rejected because ATT is implemented but the prompt is not triggered on actual devices (yet?).

Fortunately, Apple is there for us: *"More information will follow this fall."*

Did any of you experience the same madness?

---

*If you're interested, know that I've filled [a bug report](https://github.com/zoontek/react-native-permissions/issues/541) on react-native-permission's repo.*

*I doubt this is related to the plugin itself. But if it is, at least we'll know.*

**Comment by mikehardy** [comment]
@bd-arc - thanks so much for the real experience! This is helpful to me at least, and likely others.
I follow the react-native-permissions repo as well (though I don't often make PRs there) and I saw that issue. Combining the two, I think that issue - when resolved - will be the root cause and hopefully you can proceed.

**Comment by bd-arc** [comment]
@mikehardy Happy to help!

I'll share here what we've discovered in the bug report I had opened. Hopefully it will save other people a lot of headaches!

1. ⚠️  **There's a system setting that, if disabled, will prevent the ATT prompt to ever appear.**
    - See [this SO answer](https://stackoverflow.com/questions/62817164/idfa-ios14-returns-denied-without-displaying-auth-popup/62840480#62840480) for more info.
    - I personally wasn't aware of this setting's existence, both as a developer and as a user.
1. **If "disabled" is the default value of this setting, you can kiss tracking goodbye** — at least if you rely on the IDFA.
    - The odds of having someone a) finding this setting, b) voluntarily enabling it, c) allowing the tracking in your app is probably less than being hit by a meteorite.
1. Now for the funny part: **most Apple testers apparently have the system setting disabled on their own phones**. It seems crazy since they're supposed to double-check everything related to ATT, but that's the way it is.
    - We'll add some review notes to let them know that they must enable this setting before checking our app for the 6th time. I'll keep you posted on whether that worked or not.

**Comment by TaylorDale** [comment]
Is there some steps that need to be taken post-receiving a "yes" ATT response to ensure that the Admob SDK does some sort of update? Or will this all happen automatically?

**Comment by t4ngth00** [comment]
@TaylorDale I think this StackOverflow answer here https://stackoverflow.com/a/63522856 will definitely help you get your own answer.

**Comment by alinavaie** [comment]
My app has been rejected by Apple because I used both AdsConsent and AppTrackingTransparencyIOS. Do we have to use AdsConsent for android only?

**Comment by mikehardy** [comment]
I have no idea, I use neither so don't have direct experience. What do the documents upstream say? What versions of everything are you using? How is your app configured and what does a minimally reproducible example look like?

I'd ask on stackoverflow

**Comment by sdb1228** [comment]
Sorry to bring this up again even thought its closed but I wanted to follow up on @TaylorDale and @rochapablo posts [here](https://github.com/invertase/react-native-firebase/issues/4107#issuecomment-771657039) and [here](https://github.com/invertase/react-native-firebase/issues/4107#issuecomment-689514216) I read @t4ngth00 article [here](https://github.com/invertase/react-native-firebase/issues/4107#issuecomment-801876704) and am wanting some clarification around the request for permissions.  It sounds like firebase admob is built in a way that if the permission for ATT is enabled it will use IDFA and if not it will not is this correct? I understand the requesting for the permission is not the concern of firebase/admob repo, however I am curious what admob is expecting from us the developers when we receive the confirmation or denial of ATT.  I am happy to update the documentation after getting some more guidance given that this specific feature/workflow will block developer from publishing without requesting for ATT and configuring correctly with admob

**Comment by mikehardy** [comment]
Actually, admob is no longer even a part of firebase, there is no documentation to update here. Admob was removed completely upstream from firebase-android-sdk / firebase-ios-sdk / firebase-js-sdk, so there is nothing to update here with regard to admob

**Comment by sdb1228** [comment]
@mikehardy oh I am sorry is there somewhere better then I can ask this question? https://www.npmjs.com/package/@react-native-firebase/admob links to this repo and the github link here https://rnfb-docs.netlify.app/admob/european-user-consent also links to this repo.

If there is a better place for these questions to be asked or better place for documentation to be updated I am happy to go there and ask/update there as well.

**Comment by mikehardy** [comment]
Those repo links are stale, but once published you can't unpublish right? As of v12 of RNFB admob is no longer published. "rnfb-docs.netlify.app" is not an official link, rnfirebase.io is our official docs site

stackoverflow is the recommend support channel for  general firebase usage questions ("How do I... ?"), with admob tags as needed 

the upstream sdk issue trackers (firebase-ios-sdk / firebase-android-sdk) are where to post *reproductions* (based on their quickstarts provided for reproductions and API testing) of issues to address. 

For Admob I'd choose one of the three links at bottom right of Admob page under support https://developers.google.com/admob/

**Comment by sdb1228** [comment]
Do you know the owner of That netlify app so that the documentation can be updated to say deprecated? Was there an announcement somewhere that directs the users using react native firebase admob to the future of react native admob? 

Sorry if I have missed it.  I appreciate the hard work that goes into RNFB just want to make sure I am on the right migration strategy or if I need to find another package to use for react native admob.

Thanks again for your time @mikehardy

**Comment by sdb1228** [comment]
According to this link https://rnfirebase.io/migrating-to-v6 which is on the official documentation it still says that the package is located under the react native firebase.  I am a little confused if the admob package is no longer here who is hosting it? and where is the react native firebase admob package being hosted?

**Comment by mikehardy** [comment]
It's not anywhere at the moment. #5329 - Google I/O happens in May every year and brings a pile of changes. We're downstream from that so need to ingest them starting May through June, we're still backlogged on that here and naturally firebase-specific changes are first (e.g., storage emulator support, then AppCheck - in progress, firestore bundles, database.get etc). There are plans to peel admob out still and #5329 is the issue where the most traffic exists

"migrating to v6" could perhaps get a note in the admob section saying that v12 - which is about 2 years after v6 - no longer has admob so if you've waited this long to migrate off v5 the landscape is a bit different etc. I merge docs PRs almost immediately so if you hit the edit button on that one and put anything reasonable in there I can get it published

**Comment by sdb1228** [comment]
oh awesome okay I didn't see this before thanks @mikehardy 

TL;DR sounds like the wonderful people here working for free and working their hardest are getting react-native-admob migrated from RNFB admob to a new repo for future use.  If you want to help out see the links posted in @mikehardy comment above and help get some of the forward port breaking changes implemented.

Correct me if I am wrong but that is what I gleaned.  In the mean time to answer my question above I am going to try to publish with the permission check and assume that react native firebase admob is smart enough to determine if the IDFA is allowed or not.  I will report back if I see otherwise.

**Comment by sdb1228** [comment]
Update I did get approved with the ATT permissions added as noted above.  I am unsure how the integration works with admob itself, however if I notice a major impact to my revenue ill make a note here so we can further look into it.
